File: src/schema/instance.ts
import type { AssetInstanceId, PartId, TemplateId, TextureId } from './ids'
import type { TextureChannel } from './part'
import type { Transform } from './transform'
import { IDENTITY_TRANSFORM } from './transform'

export type SlotAssignment = {
  readonly partId: PartId
  readonly offset: Transform
  readonly textures: Readonly<Partial<Record<TextureChannel, TextureId>>>
}

export type AssetInstance = {
  readonly id: AssetInstanceId
  readonly name: string
  readonly templateId: TemplateId
  readonly slots: Readonly<Partial<Record<string, SlotAssignment>>>
  readonly version: number
}

export const emptyAssignment = (assignedPartId: PartId): SlotAssignment => ({
  partId: assignedPartId,
  offset: IDENTITY_TRANSFORM,
  textures: {},
})

File: src-tauri/src/schema/instance.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use ts_rs::TS;

use super::ids::{AssetInstanceId, PartId, TemplateId, TextureId};
use super::vec::Transform;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[ts(export)]
pub struct SlotAssignment {
    #[serde(rename = "partId")]
    pub part_id: PartId,
    pub offset: Transform,
    pub textures: HashMap<String, TextureId>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[ts(export)]
pub struct AssetInstance {
    pub id: AssetInstanceId,
    pub name: String,
    #[serde(rename = "templateId")]
    pub template_id: TemplateId,
    pub slots: HashMap<String, SlotAssignment>,
    pub version: u32,
}
